File: problems/linked-list/0019-remove-nth-node-from-end-of-list.js
/**
 * #19 Remove Nth Node From End of List
 * Pattern: keep two pointers n nodes apart, with a dummy node before the head.
 * Time: O(n) | Space: O(1)
 *
 * @param {ListNode|null} head
 * @param {number} n
 * @return {ListNode|null}
 */
function removeNthFromEnd(head, n) {
  const dummy = { next: head }
  let fast = dummy
  let slow = dummy

  for (let i = 0; i < n; i += 1) fast = fast.next

  while (fast.next !== null) {
    fast = fast.next
    slow = slow.next
  }

  slow.next = slow.next.next
  return dummy.next
}

module.exports = removeNthFromEnd

File: problems/linked-list/0019-remove-nth-node-from-end-of-list.py
"""#19 Remove Nth Node From End of List
Keep two pointers n links apart; a dummy node also handles removing the head.
Time: O(n) | Space: O(1)
"""


class Solution:
    def removeNthFromEnd(self, head, n):
        dummy = ListNode(0)
        dummy.next = head
        fast = slow = dummy

        for _ in range(n):
            fast = fast.next

        while fast.next is not None:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next
        return dummy.next
